tools/check_copyright.py

```diff
@@ -0,0 +1,190 @@
+an existing one.
+"""
+
+import argparse
+import datetime
+import re
+import textwrap
+
+from pathlib import Path
+
+DEFAULT_COPYRIGHT_HOLDER = 'Ekumen, Inc.'
+
+# Template for the copyright notice.
+# This should not contain any regex patterns since special characters will be escaped.
+DEFAULT_COPYRIGHT_TEMPLATE = '''\
+Copyright {year} {name}
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.\
+'''
+
+
+class FileContentChecker:
+    """Helper class to check the contents of a file and apply fixes."""
+
+    # Regex pattern that matches a year (YYYY) and a year range (YYYY-ZZZZ).
+    # It has capturing groups for both years (the second one is optional).
+    _YEAR_PATTERN = r'([0-9]{4})(?:(?:-)([0-9]{4}))?'
+
+    # Regex pattern that matches the name of a copyright holder.
+    _NAME_PATTERN = r'.*'
+
+    # Regex pattern that matches anything that looks like a copyright block.
+    # Specifically, a comment beginning with the word `copyright` (case insensitive)
+    # followed by single line comments that form a block.
+    _COPYRIGHT_PATTERN = r'(?:[;#%|*]|//)\s*[Cc]opyright.*(?:\n(?:[;#%|*]|//).*)*'
+
+    def __init__(self, path: Path):
+        """Initialize a file content checker."""
+        self._path = path
+        with open(path) as file:
+            self._content = file.read()
+
+    def _comment_lines(self, text: str) -> str:
+        """Comment text lines using the syntax associated with the file extension.
+
+        :param text: Input text.
+        :return: The modified text.
+        """
+        comment_prefix = {
+            '.py': '#',
+            '.cpp': '//',
+            '.hpp': '//',
+        }
+
+        base = textwrap.indent(text, ' ', None)
+        return textwrap.indent(
+            base, comment_prefix.get(self._path.suffix, '#'), lambda line: True
+        )
+
+    def has_copyright(self, template: str = DEFAULT_COPYRIGHT_TEMPLATE) -> bool:
+        """Verify that the file has a valid copyright notice.
+
+        :param template: Template copyright notice.
+        :return: True if a copyright notice was found. False otherwise.
+        """
+        pattern = (
+            # Comment lines and escape special characters in the template...
+            re.escape(self._comment_lines(template.format(year='YYYY', name='NAME')))
+            # Then insert year and name patterns.
+            .replace('YYYY', self._YEAR_PATTERN).replace('NAME', self._NAME_PATTERN)
```

Comment:
why combining format and replace and not directly formatting with the patterns?

Reply:
I see, thanks for clarifying. I got lost before in where each function call ended.
https://github.com/Ekumen-OS/beluga/commit/ab72795b019ed973811a2086edf16b8001e956a0 LGTM!